proposals/p1928.md

```diff
@@ -0,0 +1,163 @@
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/1928)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+
+<!-- tocstop -->
+
+## Problem
+
+Arrays are base component of almost of programming languages. In the current
+stage of Carbon development, arrays are missed.
+
+## Background
+
+An array is a set of scalar data placed in contiguous memory locations, all of
+the same type and type parameters. The scalar data that make up an array are the
+array elements which can be referenced by unique indexes.
+
+A nested array is an array of arrays which may be non-continuous in memory.
+
+A multidimensional array is an array with more than two dimensions which is
+continuous in memory.
+
+## Proposal
+
+This proposal introduces array internal representation, its declaration, and
+indexing.
+
+## Details
+
+### Internal representation
+
+For simplifying usage and raising safety of array usage, internal representation
+of arrays contains additional metadata:
+```cpp
+  struct {
+    int32_t dimensions;
+    int32_t * dimension_size;
+    void * data;
+  } Array;
+```
+Here, `dimensions` contains information about number of dimensions (necessary
+for multidimensional arrays), `dimension_size` is an array of size `dimensions`
+containing sizes for each dimensions, and `data` is continuous in memory place
+keeping scalar data.
+
+This representation is closer to `std::vector<T>` rather then `T[]`. Then,
+interoperability between C++ and carbon is limited.
+
+One-dimensional and nested arrays have `dimensions = 1`.
+
+### Declaration
+
+During declaration of arrays, it is more importantly to see type of variable.
+It is the aim of the below presented alternatives of declaration.
+
+Current realization:
+```carbon
+  var array_1D: [i32; 4];
+  var array_2D: [i32; 4, 10]; // is not supported yet
+  var array_nested: [[i32; 10]; 4];
+```
+In the case of nested array, definition of them a bit misunderstood due to
+following indexing is inversed.
+
+Suggested in [#1787](https://github.com/carbon-language/carbon-lang/pull/1787):
+```carbon
+  var array_1D: [4; i32];
+  var array_2D: [4, 10; i32];
+  var array_nested: [4; [10; i32]];
+```
+This definition is deprived of a defect of previous definition, but the type is
+the last one.
+
+Alternatives:
+```carbon
+  var array_1D[4]: i32;
+  var array_2D[4, 10]: i32;
+  var array_nested[4][10]: i32;
+```
+In this alternative, `auto` will work only for types, not for dimensions of
+array.
+
+```carbon
+  var array_1D: i32[4];
+  var array_2D: i32[4, 10];
+  var array_nested: i32[4][10];
+```
+This definition requires contextual information since `i32[4]` is the same as
+`array_1D[1]` so it is necessary to separate type and variable.
```

Comment:
I don't understand why `i32[4]` would be the same as `array_1D[1]`, under this approach.

Reply:
It seems like it might be possible to make that work, depending on how we specify `x[y]` (eg, `impl forall [N:! BigInt] Type as IndexableWith(IntLiteral(N)) where .Result == Type ...`), but it definitely seems like it'd be surprising in some cases and would probably significantly limit the semantics we can provide, such as variable-length arrays. (We might not *want* VLAs, but being unable to have them because of a syntactic choice would be concerning.)